ENH: Add macOS x86 and arm64 builds

matthewfeickert opened the pull request:
<!--
Thank you for your pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [N/A] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->

<!--
Please add any other relevant info below:
-->

matthewfeickert committed:
ENH: Add macOS x86 and arm64 builds

* Add macOS builds for x86_64 and arm64.
* Bump build number.

matthewfeickert commented:
@conda-forge-admin, please rerender

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe/meta.yaml```) and found it was in an excellent condition.

Commit:
MNT: Re-rendered with conda-smithy 3.52.3 and conda-forge-pinning 2025.10.15.11.56.55

Other tools:
- conda-build 25.9.0
- rattler-build 0.48.1
- rattler-build-conda-compat 1.4.9

matthewfeickert committed:
Decrease package size by only using required paths

kratsg commented:
looks like this additionally drops centos7, or just swaps the base image for building?

matthewfeickert committed:
try to explicitly disable python

matthewfeickert commented:
> looks like this additionally drops centos7, or just swaps the base image for building?

@kratsg that's an automatic process in the rerender, but yes, all of conda-forge is going to `alma9`.

matthewfeickert commented:
Runtime failure for x86 macOS at the example:

```
Library_Loader::LoadLibrary(): dlopen($PREFIX, 0x0009): tried: '$PREFIX' (not a file), '/System/Volumes/Preboot/Cryptexes/OS$PREFIX' (no such file), '$PREFIX' (not a file)
Library_Loader::LoadLibrary(): Failed to load library 'libLHAPDFSherpa.dylib'.
Library_Loader::LoadLibrary(): dlopen($PREFIX, 0x0009): tried: '$PREFIX' (not a file), '/System/Volumes/Preboot/Cryptexes/OS$PREFIX' (no such file), '$PREFIX' (not a file)
Library_Loader::LoadLibrary(): Failed to load library 'libLHAPDFSherpa.dylib'.
Sherpa: Variations::LoadLHAPDFInterfaceIfNecessary throws fatal error: 
   Cannot load PDF library LHAPDFSherpa
Exception_Handler::Terminate(): Pre-crash status saved to '$SRC_DIR/Examples/CI/LO_Z/Status__Tue_Dec_10_11-40-49_2024'.
```

(Same as https://github.com/conda-forge/staged-recipes/pull/28084#issuecomment-2460788212)

matthewfeickert commented on recipe/build.sh:
```diff
@@ -22,6 +22,7 @@ autoreconf --install
     --enable-hepmc2=$PREFIX \
     --enable-lhapdf=$PREFIX \
     --with-sqlite3=$PREFIX \
+    --without-python \
```
This isn't working given

```
configure: WARNING: unrecognized options: --without-python
```

kratsg replied:
sherpa is using `autconf`? I thought it had cmake support via something like `-DSHERPA_ENABLE_PYTHON=false` or something

matthewfeickert replied:
@kratsg CMake support was added in [Sherpa `v3.0.0`](https://gitlab.com/sherpa-team/sherpa/-/blob/58abdd66176f2750e5c1bfd83402bb6ebc40b3b8/CMakeLists.txt). I intentionally started with the last `v2.x` release so that if people were stuck on `v2` for whatever reason they would still have a release on conda-forge to use. `v2` is `autotools` based.

matthewfeickert commented:
```
Failed to load library 'libLHAPDFSherpa.dylib'
```

Looking in the build logs `libLHAPDFSherpa.dylib` only shows up in the line

```
libtool: install: (cd $PREFIX/lib/SHERPA-MC && { ln -s -f libLHAPDFSherpa.0.dylib libLHAPDFSherpa.dylib || { rm -f libLHAPDFSherpa.dylib && ln -s libLHAPDFSherpa.0.dylib libLHAPDFSherpa.dylib; }; })
```

but when I run the build locally with

```
python build-locally.py osx_arm64_
```

I see afterwards that only `libLHAPDFSherpa.0.dylib` is in the broken build

```
% find . -type f -iname "libLHAPDFSherpa.0.dylib"
./miniforge3/conda-bld/sherpa_1734460319284/work_moved_sherpa-2.2.16-hfe797ae_4_osx-arm64/PDF/LHAPDF/.libs/libLHAPDFSherpa.0.dylib
./miniforge3/conda-bld/sherpa_1734460319284/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/lib/SHERPA-MC/libLHAPDFSherpa.0.dylib
./miniforge3/pkgs/sherpa-2.2.16-hfe797ae_4/lib/SHERPA-MC/libLHAPDFSherpa.0.dylib
```

So for some reason `libLHAPDFSherpa.dylib` isn't getting copied over during the packaging step.

matthewfeickert committed:
Revert without-python attempt

matthewfeickert committed:
DEBUG: Add pythons for cross-compile to check

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I failed to even lint the recipe, probably because of a conda-smithy bug :cry:. This likely indicates a problem in your `meta.yaml`, though. To get a traceback to help figure out what's going on, install conda-smithy and run `conda smithy recipe-lint --conda-forge .` from the recipe directory. You can also examine the [workflow logs](https://github.com/conda-forge/conda-forge-webservices/actions/runs/12379572401) for more detail.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/12379572401](https://github.com/conda-forge/conda-forge-webservices/actions/runs/12379572401). Examine the logs at this URL for more detail.</sub>

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe/meta.yaml```) and found it was in an excellent condition.

matthewfeickert commented:
> Runtime failure for x86 macOS at the example:
> 
> ```
> Library_Loader::LoadLibrary(): dlopen($PREFIX, 0x0009): tried: '$PREFIX' (not a file), '/System/Volumes/Preboot/Cryptexes/OS$PREFIX' (no such file), '$PREFIX' (not a file)
> Library_Loader::LoadLibrary(): Failed to load library 'libLHAPDFSherpa.dylib'.
> Library_Loader::LoadLibrary(): dlopen($PREFIX, 0x0009): tried: '$PREFIX' (not a file), '/System/Volumes/Preboot/Cryptexes/OS$PREFIX' (no such file), '$PREFIX' (not a file)
> Library_Loader::LoadLibrary(): Failed to load library 'libLHAPDFSherpa.dylib'.
> Sherpa: Variations::LoadLHAPDFInterfaceIfNecessary throws fatal error: 
>    Cannot load PDF library LHAPDFSherpa
> Exception_Handler::Terminate(): Pre-crash status saved to '$SRC_DIR/Examples/CI/LO_Z/Status__Tue_Dec_10_11-40-49_2024'.
> ```
> 
> (Same as [conda-forge/staged-recipes#28084 (comment)](https://github.com/conda-forge/staged-recipes/pull/28084#issuecomment-2460788212))

`LoadLibrary("LHAPDFSherpa")` gets called only once in [`v2.2.16` SHERPA in `SHERPA/Tools/Variations.C`](https://gitlab.com/sherpa-team/sherpa/-/blob/f59a714bb661f56ecb1754bfbc734e680c8feca4/SHERPA/Tools/Variations.C#L130)

```c
  // check whether LHAPDF is already loaded, if not load and init interface
  if (!s_loader->LibraryIsLoaded("LHAPDFSherpa")) {
    s_loader->AddPath(std::string(LHAPDF_PATH)+"/lib");
    s_loader->LoadLibrary("LHAPDFSherpa");
    void *init(s_loader->GetLibraryFunction("LHAPDFSherpa","InitPDFLib"));
    if (init==NULL) THROW(fatal_error,"Cannot load PDF library LHAPDFSherpa");
    ((PDF_Init_Function)init)();
  }
```

so I assume that `LHAPDF_PATH` is getting set wrong somewhere.

**edit**: It get [set in `m4/acinclude.m4`](https://gitlab.com/sherpa-team/sherpa/-/blob/f59a714bb661f56ecb1754bfbc734e680c8feca4/m4/acinclude.m4#L771)

```configure
    AC_DEFINE_UNQUOTED([LHAPDF_PATH], "$CONDITIONAL_LHAPDFDIR", [LHAPDF directory])
```

matthewfeickert commented:
@chrisburr @kratsg not sure if you have any ideas here. I haven't dug very deep into it yet, but it is curious to me (maybe naively so) that this only seems to be affecting macOS.

I'm also going to abuse knowing @20DM's GitHub handle to ping them as well, given that they're a Sherpa author.

kratsg commented:
> @chrisburr @kratsg not sure if you have any ideas here. I haven't dug very deep into it yet, but it is curious to me (maybe naively so) that this only seems to be affecting macOS.

I can't think of anything that would force this to be Mac-OSX specific unless there's some darwin flag that tells sherpa to rely on the python that ships with Mac (which isn't really the case for future OS releases).

matthewfeickert commented:
> I can't think of anything that would force this to be Mac-OSX specific unless there's some darwin flag that tells sherpa to rely on the python that ships with Mac (which isn't really the case for future OS releases).

AFAIK, the

```
Library_Loader::LoadLibrary(): dlopen($PREFIX, 0x0009): tried: '$PREFIX' (not a file), '/System/Volumes/Preboot/Cryptexes/OS$PREFIX' (no such file), '$PREFIX' (not a file)
Library_Loader::LoadLibrary(): Failed to load library 'libLHAPDFSherpa.dylib'.
```

issue isn't Python related.

matthewfeickert commented:
Something seems to be going wrong in the packaging step, as I can create a minimal `pixi` project on my arm64 macOS machine that will build from `v2.2.16` source with the same patches applied, install, and then be able to run the example with no errors.

That or there's a problem with [`ATOOLS/Org/Library_Loader.C`'s `Library_Loader::LoadLibrary`](https://gitlab.com/sherpa-team/sherpa/-/blob/f59a714bb661f56ecb1754bfbc734e680c8feca4/ATOOLS/Org/Library_Loader.C#L82-L120) which is complaining about `$PREFIX`.

**edit**: Yeah, something is going wrong in packaging, as if I put in a

```
$(command -v Sherpa) -f ./Examples/CI/LO_Z/Run.dat
```

just after `make install` in the `build.sh`, **that** example will pass. But the same example will fail in the `test` stage post packaging.
